### ISSUE TITLE
func-style like plugin

### ISSUE THREAD

**Issue by jamestalmage** [sindresorhus/eslint-plugin-xo#3]
See: http://eslint.org/docs/rules/func-style

My personal preference is to prefer `declaration` style, as it allows hoisting (and just looks better to me). The only issue is when you want to reassign:

```js
var func = function () {

}

if (someCondition) {
  func = wrap(func);
}
```

ESLint's `func-style` rule does not accommodate the above exception when preferring declarations, so it is impractical to enforce. 

---

Note: 

You can get around ELint's rule by doing this:

```js
var func; // don't initialize variable here, do it on the next line

func = function () {

}
```

But that feels impractical.

---

Do we feel it's worth enforcing `func-style` in XO? Should we use the builtin plugin to do so, or make our own that is a bit more permissive.

**Comment by jfmengels** [comment]
When writing in ES5, I prefer the `declaration` style, when writing with ES2015, I think I preferred (has been a while now) writing `expression` style for short functions (`const double = (x) => x * 2;`), and `declaration` style for longer ones.  With the `expression` one, you have the benefit that you know that the function won't be re-declared elsewhere if you've assigned it to a `const` variable (and I don't like hoisting btw).

So I'm rather against enforcing it.

**Comment by jamestalmage** [comment]
Arrow functions would be excluded (that's an option even in ESLints plugin).

**Comment by jfmengels** [comment]
With arrow function excluded... then maybe yeah.
We'll still have to mix decorated stuff like this anyway:
```js
var fn = _.curry(function (...) { ... });
```
But yeah, why not.

**Comment by jamestalmage** [comment]
The eslint rule (in `declaration` mode, with arrow functions excluded), only flags a `FunctionExpression` if it's immediate parent is a `VariableDeclarator`, so you only run afoul doing `var fn = function () {}`. 

The ESLint behavior makes sense to me, but it doesn't make an exception for variables that are later reassigned. That is a rare occurrence, but if you are already in the habit of using function declarations, then `var fn = function () {}` is a telegraphs the potential reassignment to the reader.

Jumping back to arrow functions for a second:

I think they should only be allowed for simple functions with implicit returns

```js
// OK
var square = x => x * x;

// NOT OK
var complexFunc = (x, y) => {
  if (x) {
    return y * 2;
  } else { 
    return y * 3;
  }
}

// I would prefer
function complexFunc(x, y) {
  if (x) {
    return y * 2;
  } else { 
    return y * 3;
  }
}
```

Exceptions for arrow funcs should be made if they use any of `this`, `arguments`, `super` or `new.target`.

**Comment by jamestalmage** [comment]
A bit dated, but: https://kangax.github.io/nfe/

**Comment by sindresorhus** [comment]
> I think they should only be allowed for simple functions with implicit returns

Why? I prefer arrow functions for everything as they're lexically scoped (no dumb implicit `this`) and you can use `const` with them to ensure they're never overridden.

**Comment by jamestalmage** [comment]
Well, if you don't need the lexical binding there's no advantage. Actually there may be a performance hit if arrow functions are transpiled (to accommodate the binding, etc) - though that might not even true - not sure if Babel is smart enough to avoid binding if it isn't needed.

Still, I am not dogmatic about it.

This came up because I noticed we're mixing styles across the AVA codebase, and was wondering if there's a good set of rules we can use to describe our preferred style (whatever that is).

**Comment by sindresorhus** [comment]
Yes, I like to figure out a good default we can enforce for this too. I don't feel super strongly either way, just slightly prefer arrow functions. Also nested function declarations looks messy and having everything as variables looks cleaner and more consistent.

// @vdemedes @novemberborn @SamVerschueren

> if you don't need the lexical binding there's no advantage.

Still the advantage of `const` and no hoisting. I consider hoisting a misfeature. I prefer explicitness.

> Actually there may be a performance hit if arrow functions are transpiled

Node.js 4 supports arrow functions though. So I don't see it as a big problem going forward.

**Comment by novemberborn** [comment]
I tend to use a mix of syntaxes. E.g. in [`package-hash`](https://github.com/novemberborn/package-hash/blob/69a328840818d336ae13af22f9af092208fcb49b/src/index.js), because I'm using `export function sync` I like declaration style for the other functions in the file.

In other files where there are few declarations I'm [more likely](https://github.com/novemberborn/copenhagen/blob/5ff8739b17368106570bc1551eb131625a6fb63d/test/coverage.js#L14) to use `const` with an arrow function, but I'm [far from consistent](https://github.com/novemberborn/copenhagen/blob/5ff8739b17368106570bc1551eb131625a6fb63d/test/add-instrumentation.js#L11) about this.

If I'd have to choose I'd prefer `const` (and in rare occasions `let`) with arrow functions. Never `var`. But not for exports… I'm not sure there's a single consistent answer here.

**Comment by sindresorhus** [comment]
I agree that ES2015 exports should use a normal function.

This:

```js
export function foo() {}
```

Looks better than this:

```js
export const foo = () => {}
```

**Comment by jacekkopecky** [comment]
Related to this (old-ish) issue, I've recently implemented and submitted related PR to eslint, not realizing it had no chance of being accepted: [eslint#16159](https://github.com/eslint/eslint/pull/16159) . 

I too prefer the declaration style, with the exception where reassignment is desired, where fat arrow functions look better to me. The current `func-style` rule in eslint does not allow the exception, my PR does. Shall I repurpose the PR for plugin-unicorn?
